ably/options.go

```diff
@@ -292,56 +338,87 @@ func (opts *ClientOptions) disconnectedRetryTimeout() time.Duration {
 }
 
-func (opts *ClientOptions) restURL() string {
+func (opts *ClientOptions) getRestHost() string {
 	host := opts.RestHost
-	if host == "" {
-		host = defaultOptions.RestHost
-		if opts.Environment == "production" {
-			opts.Environment = ""
-		}
-		if opts.Environment != "" {
-			host = opts.Environment + "-" + host
-		}
+	defaultHost := defaultOptions.RestHost
+	if empty(host) {
+		host = defaultHost
 	}
-	if opts.NoTLS {
-		port := opts.Port
-		if port == 0 {
-			port = 80
-		}
-		return "http://" + net.JoinHostPort(host, strconv.FormatInt(int64(port), 10))
-	} else {
-		port := opts.TLSPort
-		if port == 0 {
-			port = 443
-		}
-		return "https://" + net.JoinHostPort(host, strconv.FormatInt(int64(port), 10))
+	if host == defaultHost && !opts.isProductionEnvironment() {
+		host = opts.Environment + "-" + host
 	}
+	return host
 }
 
-func (opts *ClientOptions) realtimeURL() string {
+func (opts *ClientOptions) getRealtimeHost() string {
 	host := opts.RealtimeHost
-	if host == "" {
-		host = defaultOptions.RealtimeHost
-		if opts.Environment == "production" {
-			opts.Environment = ""
-		}
-		if opts.Environment != "" {
-			host = opts.Environment + "-" + host
+	defaultHost := defaultOptions.RealtimeHost
+	if empty(host) {
+		if !empty(opts.RestHost) {
+			logger := opts.Logger.Sugar()
+			logger.Warnf(`restHost is set to %s but realtimeHost is not set so setting realtimeHost to %s too. 
+							If this is not what you want, please set realtimeHost explicitly.`, opts.RestHost, opts.RealtimeHost)
+			return opts.RestHost
 		}
+		host = defaultHost
 	}
+	if host == defaultHost && !opts.isProductionEnvironment() {
+		host = opts.Environment + "-" + host
+	}
+	return host
+}
+
+func empty(s string) bool {
+	return len(strings.TrimSpace(s)) == 0
+}
+
+func (opts *ClientOptions) restURL() (restUrl string) {
+	restHost := opts.getRestHost()
+	port, _ := opts.activePort()
+	baseUrl := net.JoinHostPort(restHost, strconv.FormatInt(int64(port), 10))
 	if opts.NoTLS {
-		port := opts.Port
-		if port == 0 {
-			port = 80
+		return "http://" + baseUrl
+	}
+	return "https://" + baseUrl
+}
+
+func (opts *ClientOptions) realtimeURL() (realtimeUrl string) {
+	realtimeHost := opts.getRealtimeHost()
+	port, _ := opts.activePort()
+	baseUrl := net.JoinHostPort(realtimeHost, strconv.FormatInt(int64(port), 10))
+	if opts.NoTLS {
+		return "ws://" + baseUrl
+	}
+	return "wss://" + baseUrl
+}
+
+func (opts *ClientOptions) getFallbackHosts() ([]string, error) {
+	logger := opts.Logger.Sugar()
+	_, isDefaultPort := opts.activePort()
+	if opts.FallbackHostsUseDefault {
+		if opts.FallbackHosts != nil {
+			return nil, fmt.Errorf("fallbackHosts and fallbackHostsUseDefault cannot both be set")
 		}
-		return "ws://" + net.JoinHostPort(host, strconv.FormatInt(int64(port), 10))
-	} else {
-		port := opts.TLSPort
-		if port == 0 {
-			port = 443
+		if !isDefaultPort {
+			return nil, fmt.Errorf("fallbackHostsUseDefault cannot be set when port or tlsPort are set")
```

Comment:
Use [`errors.New`](https://golang.org/pkg/errors/#New) rather than `fmt.Errorf` (since nothing is being formatted)